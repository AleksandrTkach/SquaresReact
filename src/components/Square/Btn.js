import React from 'react';
import PropTypes from 'prop-types';

export default class Btn extends React.Component {
	_getValues = (type, posBtnMinusRow, posBtnMinusCol, paddingSize) => {
		let classNameDiv;
		let classNameIcon;
		let stylePosition;

		switch (type) {
			case 'plus-col':
				classNameDiv = 'tile_orange tile_plus-col';
				classNameIcon = 'plus';
				break;
			case 'plus-row':
				classNameDiv = 'tile_orange tile_plus-row';
				classNameIcon = 'plus';
				stylePosition = {
					marginLeft: `${paddingSize}px`,
				};
				break;
			case 'minus-row':
				classNameDiv = 'tile_red tile_minus-row';
				classNameIcon = 'minus';
				stylePosition = {
					top: `${posBtnMinusRow}px`,
				};
				break;
			case 'minus-col':
				classNameDiv = 'tile_red tile_minus-col';
				classNameIcon = 'minus';
				stylePosition = {
					left: `${posBtnMinusCol}px`,
				};
				break;
			default:
				classNameDiv = '';
				classNameIcon = '';
				stylePosition = '';
		}

		return {
			classNameDiv,
			classNameIcon,
			stylePosition,
		};
	};

	render() {
		const {
			cellSize,
			type,
			actionOnClick,
			posBtnMinusRow,
			posBtnMinusCol,
			paddingSize,
		} = this.props;

		const styleSize = {
			height: `${cellSize}px`,
			width: `${cellSize}px`,
		};

		const { classNameDiv, stylePosition, classNameIcon } = this._getValues(
			type,
			posBtnMinusRow,
			posBtnMinusCol,
			paddingSize
		);

		return (
			<div
				id={`btn-${type}`}
				className={`tile ${classNameDiv}`}
				style={{ ...styleSize, ...stylePosition }}
				onClick={actionOnClick}
			>
				<i className={`fa fa-${classNameIcon}`} />
			</div>
		);
	}
}

Btn.propTypes = {
	cellSize: PropTypes.number.isRequired,
	type: PropTypes.string.isRequired,
	actionOnClick: PropTypes.func.isRequired,
	paddingSize: PropTypes.number,
	posBtnMinusRow: PropTypes.number,
	posBtnMinusCol: PropTypes.number,
};
