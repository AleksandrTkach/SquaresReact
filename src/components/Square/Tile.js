import React from 'react';
import PropTypes from 'prop-types';

export default class Tile extends React.Component {
	render() {
		const { cellSize, checkPosition, indexRow, indexCol } = this.props;

		const style = {
			height: `${cellSize}px`,
			width: `${cellSize}px`,
		};

		return (
			<div
				className="tile tile_blue"
				style={style}
				onMouseOver={() => checkPosition(indexRow, indexCol)}
			/>
		);
	}
}

Tile.propTypes = {
	cellSize: PropTypes.number.isRequired,
	indexRow: PropTypes.number.isRequired,
	indexCol: PropTypes.number.isRequired,
};
