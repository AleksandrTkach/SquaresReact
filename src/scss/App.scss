html,
body {
  margin: 0;
  padding: 0;
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100%;
}

.square {
  position: relative;
  display: inline-block;
  border: 2px solid deepskyblue;

  &__line {
    display: flex;
  }

  &:hover .block_minus-column{
    visibility: visible;
    opacity: 1;
    right: 100%;
  }
}

.block {
  display: flex;
  justify-content: center;
  align-items: center;
  color: white;

  &__blue {
    background-color: deepskyblue;
    border: 1px solid white;
  }
  &__orange {
    background-color: orange;
    cursor: pointer;
  }
  &__red {
    background-color: darkred;
    cursor: pointer;
  }

  &_plus {
    &-row {
      position: absolute;
      top: 100%;
      margin: 3px 0;
    }
    &-column {
      position: absolute;
      top: 0;
      left: 100%;
      margin: 0 3px;
    }
  }

  &_minus {
    &-row,
    &-column {
      visibility: hidden;
      opacity: 0;
      position: absolute;
      transition: .25s;
    }

    &-row {
      bottom: 100%;
      margin: 3px 0;
    }

    &-column {
      top: 0;
      right: 50%;
      margin: 0 3px;
      z-index: -1;
    }
  }
}
