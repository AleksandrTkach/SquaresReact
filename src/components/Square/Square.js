import React, { Component } from 'react';
import Tile from './Tile';
import Btn from './Btn';

const PADDING_SIZE = Number(process.env.REACT_APP_PADDING_SIZE);

export default class Square extends Component {
	constructor(props) {
		super(props);

		const { initialHeight, initialWidth, cellSize } = this.props;

		this.state = {
			square: [],
			initialHeight: initialHeight,
			initialWidth: initialWidth,
			cellSize,
			posCol: 0,
			posRow: 0,
			iCol: 0,
			iRow: 0,
			counter: 0,
		};
	}

	componentDidMount() {
		this._buildSquare();
	}

	/**
	 * Check position btns minus
	 *
	 * @param iRow
	 * @param iCol
	 * @param cellSize
	 */
	checkPosition = (
		iRow = this.state.iRow,
		iCol = this.state.iCol,
		cellSize = this.state.cellSize
	) => {
		console.log(iRow, ' : ', iCol);
		this.setState({
			posCol: cellSize * iCol + PADDING_SIZE * iCol,
			posRow: cellSize * iRow + PADDING_SIZE * iRow,
			iCol,
			iRow,
		});
	};

	/**
	 * Build square
	 *
	 * @private
	 */
	_buildSquare = () => {
		const { initialHeight, initialWidth } = this.state;

		let square = [];
		let counter = 0;

		for (let row = 0; row < initialHeight; row++) {
			let tiles = [];

			for (let col = 0; col < initialWidth; col++) {
				tiles[col] = {
					id: counter++,
				};
			}
			square[row] = {
				id: counter++,
				tiles: tiles,
			};
		}
		console.log('square: ', square);
		this.setState({
			square,
		});
	};

	_setTile = (row, col, cellSize = this.state.cellSize) => (
		<Tile
			key={`${row}-${col}`}
			cellSize={cellSize}
			row={row}
			col={col}
			checkPosition={this.checkPosition}
		/>
	);

	_addCol = () => {
		const { initialWidth: col } = this.state;
		let square = [...this.state.square];

		square.map((row, index) => {
			row.tiles.push(this._setTile(index, col));
			return row;
		});

		this.setState({
			square,
		});
	};

	_addRow = () => {
		const { initialHeight: row, initialWidth } = this.state;

		let square = [...this.state.square];
		let tiles = [];

		for (let col = 0; col < initialWidth; col++) {
			tiles.push(this._setTile(row, col));
		}

		square[row] = {
			tiles,
		};

		this.setState({
			square,
		});
	};

	_removeCol = async () => {
		const { iCol } = this.state;
		let square = [...this.state.square];

		square.map(row => row.tiles.filter((tile, index) => index !== iCol));

		await this.setState({
			square,
		});

		this._buildSquare();
	};

	_removeRow = async () => {
		const { square, iRow } = this.state;

		await this.setState({
			square: square.filter((row, index) => index !== iRow),
		});

		this._buildSquare();
	};

	render() {
		const {
			square,
			posCol,
			posRow,
			initialWidth,
			initialHeight,
			cellSize,
		} = this.state;

		const btnRemoveCol =
			initialWidth > 1 ? (
				<Btn
					cellSize={cellSize}
					type="minus-row"
					onClick={this._removeRow}
					posCol={posCol}
					posRow={posRow}
					paddingSize={PADDING_SIZE}
				/>
			) : (
				''
			);
		const btnRemoveRow =
			initialHeight > 1 ? (
				<Btn
					cellSize={cellSize}
					type="minus-col"
					onClick={this._removeCol}
					posCol={posCol}
					posRow={posRow}
					paddingSize={PADDING_SIZE}
				/>
			) : (
				''
			);

		return (
			<div className="square">
				<div className="tiles__wrapper">
					{square.map((row, iRow) => (
						<div key={`row-${row.id}`} className="row">
							{row.tiles.map((tile, iCol) => (
								<Tile
									key={`${tile.id}`}
									cellSize={cellSize}
									iRow={iRow}
									iCol={iCol}
									checkPosition={this.checkPosition}
								/>
							))}
						</div>
					))}

					{btnRemoveCol}
					{btnRemoveRow}
				</div>

				<Btn
					cellSize={cellSize}
					type="plus-row"
					onClick={this._addRow}
					paddingSize={PADDING_SIZE}
				/>
				<Btn cellSize={cellSize} type="plus-col" onClick={this._addCol} />
			</div>
		);
	}
}
