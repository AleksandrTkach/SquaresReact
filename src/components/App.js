import React, {useState} from 'react';
import Square from './Square/Square';
import 'scss/App.scss';

const INITIAL_HEIGHT = Number(process.env.REACT_APP_INITIAL_HEIGHT);
const INITIAL_WIDTH = Number(process.env.REACT_APP_INITIAL_WIDTH);
const CELL_SIZE = process.env.REACT_APP_CELL_SIZE;

const styleBlock = {
    height: `${CELL_SIZE}px`,
    width: `${CELL_SIZE}px`,
};

export const App = () => {

    const [squareHeight, setHeight] = useState(INITIAL_HEIGHT);
    const [squareWidth, setWidth] = useState(INITIAL_WIDTH);

    return (
      <>
        <div className="square">
          <Square styleBlock={styleBlock} initialHeight={squareHeight} initialWidth={squareWidth} />
          <div className="block block__orange block_plus-row" style={styleBlock} onClick={() => setHeight(squareHeight + 1)}>
              <i className="fa fa-plus"/></div>
          <div className="block block__orange block_plus-column" style={styleBlock} onClick={() => setWidth(squareWidth + 1)}>
              <i className="fa fa-plus"/></div>
          <div className="block block__red block_minus-column" style={styleBlock}> <i className="fa fa-minus"/> </div>
        </div>
      </>
    );
}

export default App
