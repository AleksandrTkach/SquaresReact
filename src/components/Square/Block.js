import React, {useContext} from 'react';
import {PositionBtnMinusContent} from 'components/App';
import {PaddingSize} from 'scss/variables/Variables.js';

const cellSize = process.env.REACT_APP_CELL_SIZE;

const Block = ({iRow, iCol, styleBlockSize}) => {
    const {setPosition} = useContext(PositionBtnMinusContent);

    const _checkPosition = () => {
        setPosition({
            'posRow': iRow * cellSize + PaddingSize * iRow,
            'posCol': iCol * cellSize + PaddingSize * iCol
        });

    };

    return (
      <div
        className="block block__blue"
        style={styleBlockSize}
        onMouseOver={_checkPosition}
        onFocus={_checkPosition}
      />
    )
};

export default Block

