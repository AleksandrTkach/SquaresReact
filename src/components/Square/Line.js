import React from 'react';
import Block from './Block';

export default function Line({styleBlock, initialWidth, columnI}) {
    const line = [];
    for (let i = 0; i < initialWidth; i++) {
        line.push(<Block key={i} columnI={columnI} lineI={i} styleBlock={styleBlock} />);
    }

    return(
      <div className="square__line">
        {line}
      </div>
    )
}


