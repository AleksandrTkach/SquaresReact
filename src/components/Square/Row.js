import React from 'react';
import Block from './Block';

export default function Row({styleBlock, initialWidth, columnI}) {
    const row = [];
    for (let i = 0; i < initialWidth; i++) {
        row.push(<Block key={i} columnI={columnI} lineI={i} styleBlock={styleBlock} />);
    }

    return(
      <div className="square__line">
        {row}
      </div>
    )
}


