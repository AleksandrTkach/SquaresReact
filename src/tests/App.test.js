import React from 'react';
import { shallow, mount } from 'enzyme';

import App from 'components/App';
import Square from 'components/Square/Square';
import Tile from 'components/Square/Tile';

const currentHeight = () => SquareMount.find('.row').length;
const currentWidth = () => SquareMount.find('.row').at(0).find(Tile).length;

const AppShallow = shallow(<App />);
const SquareShallow = shallow(<Square />);
const SquareMount = mount(<Square />);
const TileShallow = shallow(<Tile indexCol={1} indexRow={1} cellSize={50}/>);

describe('Render', () => {
	it('App', () => expect(AppShallow).toMatchSnapshot());
	it('Square', () => expect(SquareShallow).toMatchSnapshot());
	it('Tile', () => expect(TileShallow).toMatchSnapshot());
});

describe('Square', () => {

	it('initialHeight', () => {
		const {initialHeight} = AppShallow.props();
		expect(SquareShallow.find('.row')).toHaveLength(initialHeight);
	});

	it('initialWidth', () => {
		const {initialWidth} = AppShallow.props();
		expect(SquareShallow.find('.row').at(0).find(Tile)).toHaveLength(initialWidth);
	});

	it('plusRow', () => {
		const squareHeight = currentHeight();
		SquareMount.find('#btn-plus-row').simulate('click');
		expect(SquareMount.find('.row')).toHaveLength(squareHeight + 1);
	});

	it('plusCol', () => {
		const squareWidth = currentWidth();
		SquareMount.find('#btn-plus-col').simulate('click');
		expect(SquareMount.find('.row').at(0).find(Tile)).toHaveLength(squareWidth + 1);
	});

	it('minusRow', () => {
		const squareHeight = currentHeight();
		SquareMount.find('#btn-minus-row').simulate('click');
		expect(SquareMount.find('.row')).toHaveLength(squareHeight - 1);
	});

	it('minusCol', () => {
		const squareWidth = currentWidth();
		SquareMount.find('#btn-minus-col').simulate('click');
		expect(SquareMount.find('.row').at(0).find(Tile)).toHaveLength(squareWidth - 1);
	});

	it('lastRowNotRemoved', () => {
		const SquareMount = mount(<Square initialHeight={1} initialWidth={1} cellSize={1}/>);
		SquareMount.find('#btn-minus-row').simulate('click');
		expect(SquareMount.find('.row')).toHaveLength(1);
	});

	it('lastColNotRemoved', () => {
		const SquareMount = mount(<Square initialHeight={1} initialWidth={1} cellSize={1}/>);
		SquareMount.find('#btn-minus-col').simulate('click');
		expect(SquareMount.find('.row').at(0).find(Tile)).toHaveLength(1);
	});
});
