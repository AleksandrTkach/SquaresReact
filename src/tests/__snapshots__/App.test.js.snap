// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render App 1`] = `
<Square
  cellSize={50}
  initialHeight={4}
  initialWidth={4}
/>
`;

exports[`Render Square 1`] = `
<div
  className="square"
  style={
    Object {
      "margin": "52px",
    }
  }
>
  <div
    className="tiles__wrapper"
  >
    <div
      className="row"
      key="row-4"
    >
      <Tile
        cellSize={50}
        checkPosition={[Function]}
        indexCol={0}
        indexRow={0}
        key="tile-0"
      />
      <Tile
        cellSize={50}
        checkPosition={[Function]}
        indexCol={1}
        indexRow={0}
        key="tile-1"
      />
      <Tile
        cellSize={50}
        checkPosition={[Function]}
        indexCol={2}
        indexRow={0}
        key="tile-2"
      />
      <Tile
        cellSize={50}
        checkPosition={[Function]}
        indexCol={3}
        indexRow={0}
        key="tile-3"
      />
    </div>
    <div
      className="row"
      key="row-9"
    >
      <Tile
        cellSize={50}
        checkPosition={[Function]}
        indexCol={0}
        indexRow={1}
        key="tile-5"
      />
      <Tile
        cellSize={50}
        checkPosition={[Function]}
        indexCol={1}
        indexRow={1}
        key="tile-6"
      />
      <Tile
        cellSize={50}
        checkPosition={[Function]}
        indexCol={2}
        indexRow={1}
        key="tile-7"
      />
      <Tile
        cellSize={50}
        checkPosition={[Function]}
        indexCol={3}
        indexRow={1}
        key="tile-8"
      />
    </div>
    <div
      className="row"
      key="row-14"
    >
      <Tile
        cellSize={50}
        checkPosition={[Function]}
        indexCol={0}
        indexRow={2}
        key="tile-10"
      />
      <Tile
        cellSize={50}
        checkPosition={[Function]}
        indexCol={1}
        indexRow={2}
        key="tile-11"
      />
      <Tile
        cellSize={50}
        checkPosition={[Function]}
        indexCol={2}
        indexRow={2}
        key="tile-12"
      />
      <Tile
        cellSize={50}
        checkPosition={[Function]}
        indexCol={3}
        indexRow={2}
        key="tile-13"
      />
    </div>
    <div
      className="row"
      key="row-19"
    >
      <Tile
        cellSize={50}
        checkPosition={[Function]}
        indexCol={0}
        indexRow={3}
        key="tile-15"
      />
      <Tile
        cellSize={50}
        checkPosition={[Function]}
        indexCol={1}
        indexRow={3}
        key="tile-16"
      />
      <Tile
        cellSize={50}
        checkPosition={[Function]}
        indexCol={2}
        indexRow={3}
        key="tile-17"
      />
      <Tile
        cellSize={50}
        checkPosition={[Function]}
        indexCol={3}
        indexRow={3}
        key="tile-18"
      />
    </div>
    <Btn
      actionOnClick={[Function]}
      cellSize={50}
      paddingSize={2}
      posBtnMinusCol={0}
      posBtnMinusRow={0}
      type="minus-row"
    />
    <Btn
      actionOnClick={[Function]}
      cellSize={50}
      paddingSize={2}
      posBtnMinusCol={0}
      posBtnMinusRow={0}
      type="minus-col"
    />
  </div>
  <Btn
    actionOnClick={[Function]}
    cellSize={50}
    paddingSize={2}
    type="plus-row"
  />
  <Btn
    actionOnClick={[Function]}
    cellSize={50}
    type="plus-col"
  />
</div>
`;

exports[`Render Tile 1`] = `
<div
  className="tile tile_blue"
  onMouseOver={[Function]}
  style={
    Object {
      "height": "50px",
      "width": "50px",
    }
  }
/>
`;
